docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,732 @@
+other than Rust, this leads the user to believe that if an instruction
+compiles, it will run. There are multiple ways of achieving the same
+result that do not force the user to have separated (sometimes by
+multiple hundreds of lines of code) implementations of the forward and
+inverse operations.
+
+Furthermore, because =Expressions= are used mainly as arguments, and
+not as general purpose-combination devices, the same devices need to
+be replicated in the instruction set as =InstructionBox= variants.
+
+We must now talk about the instruction application procedure. The
+instructions can have multiple side-effects. As such, because the
+instructions represent atomic[fn:1] =WorldStateView= modifications, if a
+particular instruction in a transaction fails to apply, all
+instructions effects must be reversed. But in the given model, such a
+concept might not be possible to define, and as such, the only way to
+enforce atomicity, given that any transaction can affect potentially
+an unconstrained number of objects, the result is that the largest
+object in the blockchain must be cloned before each application. There
+are relatively efficient ways of mitigating this with persistent data
+structures but all of them pale in comparison with the efficiency of
+having a simpler instruction set in which a tiny addition to a number
+is reverted by changing that tiny number back. This is not impossible
+to achieve with the current architecture, but it would result in many
+redundant pieces of code.
+
+There is also the matter of mixing levels of abstraction. =Mint= is a
+financial operation, =If= is logical, =FailBox=, probably represents
+an instruction that fails always, but why does it have the word =Box=
+in it? More importantly by being specific we lose flexibility to allow
+different kinds of blockchains to cohesively specialise. It is poorly
+defined whether a =Transfer= would allow exchange of currencies in the
+blockchain, and with instructions being defined host-side and having
+no other mechanism of expressing the concepts, we are in the
+precarious position of not being able to leave this up to the
+individual blockchains. Further, the entire set of =Grant= and
+=Revoke= instructions is made useless in the public use case, nudging
+developers of public blockchains to use this functionality as it
+happened with =Metadata=.
+
+In addition to conceptual redundancy, there is also redundancy between
+different elements of the ISA. We have both a =SequenceBox=, and
+multiple instructions comprising a single transaction. There's both an
+=If= as an =instruction= and an =If= as an =Expression=. Further
+obfuscating the types, each instruction accepts an expression which
+either wraps a =Value= or a collection of operations that are
+ill-defined for some of the potential variants of =Value=.
+
+Finally, compounding the issue, the Iroha special instructions were
+meant to be accompanied by a domain-specific language that is only now
+being written.
```

Comment:
Were they really meant to be used in DSL? I haven't seen such an issue/task, while I've seen a lot of overs old ones

Reply:
There was a `dsl` crate that I removed in November 2021. The intentions were not always committed to GitHub, because we used to use Jira for the issue tracking until August 2021.